server.go

```diff
@@ -495,45 +499,56 @@ func (srv *Server) serveUDP(l net.PacketConn) error {
 	}
 
 	var wg sync.WaitGroup
-	defer func() {
-		wg.Wait()
-		close(srv.shutdown)
-	}()
 
 	rtimeout := srv.getReadTimeout()
-	// deadline is not used here
-	for srv.isStarted() {
-		var (
-			m    []byte
-			sPC  net.Addr
-			sUDP *SessionUDP
-			err  error
-		)
-		if isUDP {
-			m, sUDP, err = reader.ReadUDP(lUDP, rtimeout)
-		} else {
-			m, sPC, err = readerPC.ReadPacketConn(l, rtimeout)
-		}
-		if err != nil {
-			if !srv.isStarted() {
-				return nil
-			}
-			if netErr, ok := err.(net.Error); ok && netErr.Temporary() {
-				continue
-			}
-			return err
-		}
-		if len(m) < headerSize {
-			if cap(m) == srv.UDPSize {
-				srv.udpPool.Put(m[:srv.UDPSize])
+
+	g := new(errgroup.Group)
+
+	for i := 0; i < runtime.NumCPU(); i++ {
+		g.Go(func() error {
+			// deadline is not used here
+			for srv.isStarted() {
+				var (
+					m    []byte
+					sPC  net.Addr
+					sUDP *SessionUDP
+					err  error
+				)
+
+				if isUDP {
+					m, sUDP, err = reader.ReadUDP(lUDP, rtimeout)
```

Comment:
There's nothing to actually ensure this call unblocks when the errgroup returns with an error. This could easily end up still reading in one go routine while others have already exited.